compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerToUKernels.cpp

```diff
@@ -152,49 +144,199 @@ static FailureOr<IREE::Codegen::UKernelOpInterface> matchDAGForUKernel(
+    RewriterBase &rewriter, tensor::UnPackOp op) {
+  Value in = op.getSource();
+  Value out = op.getDest();
+  auto inType = in.getType().cast<ShapedType>();
+  auto outType = out.getType().cast<ShapedType>();
+  Type inElemType = inType.getElementType();
+  Type outElemType = outType.getElementType();
+  std::string fnName;
+  if (inElemType.isSignlessInteger(32) && outElemType.isSignlessInteger(32)) {
+    fnName = "vmvx.unpack.i32i32";
+  } else if (inElemType.isF32() && outElemType.isF32()) {
+    fnName = "vmvx.unpack.f32f32";
+  } else {
+    return rewriter.notifyMatchFailure(
+        op, "unsupported combination of element types");
+  }
+
+  if (inType.getRank() != 4) {
+    return rewriter.notifyMatchFailure(op, "expected input to be 4D");
+  }
+
+  if (outType.getRank() != 2) {
+    return rewriter.notifyMatchFailure(op, "expected output to be 2D");
+  }
+
+  int64_t innerDimsPos[2] = {0, 1};
+  ArrayRef<int64_t> innerDimsPosArr = op.getInnerDimsPos();
+  if (!innerDimsPosArr.empty()) {
+    innerDimsPos[0] = innerDimsPosArr[0];
+    innerDimsPos[1] = innerDimsPosArr[1];
+  }
+
+  int64_t outerDimsPerm[2] = {0, 1};
+  ArrayRef<int64_t> outerDimsPosArr = op.getOuterDimsPerm();
+  if (!outerDimsPosArr.empty()) {
+    outerDimsPerm[0] = outerDimsPosArr[0];
+    outerDimsPerm[1] = outerDimsPosArr[1];
+  }
+
+  int flags = 0;
+
+  if (innerDimsPos[0] == 0 && innerDimsPos[1] == 1) {
+    // nothing to do
+  } else if (innerDimsPos[0] == 1 && innerDimsPos[1] == 0) {
+    flags |= IREE_UK_FLAG_UNPACK_TRANSPOSE_INNER;
+  } else {
+    return rewriter.notifyMatchFailure(op, "unsupported inner_dims_pos");
+  }
+
+  if (outerDimsPerm[0] == 0 && outerDimsPerm[1] == 1) {
+    // nothing to do
+  } else if (outerDimsPerm[0] == 1 && outerDimsPerm[1] == 0) {
+    flags |= IREE_UK_FLAG_UNPACK_TRANSPOSE_OUTER;
+  } else {
+    return rewriter.notifyMatchFailure(op, "unsupported outer_dims_perm");
+  }
+
+  Location loc = op.getLoc();
+  Value in_size0 = rewriter.create<tensor::DimOp>(loc, in, 0);
+  Value in_size1 = rewriter.create<tensor::DimOp>(loc, in, 1);
+  Value in_size2 = rewriter.create<tensor::DimOp>(loc, in, 2);
+  Value in_size3 = rewriter.create<tensor::DimOp>(loc, in, 3);
+  Value out_size0 = rewriter.create<tensor::DimOp>(loc, out, 0);
+  Value out_size1 = rewriter.create<tensor::DimOp>(loc, out, 1);
+  Value flagsVal = rewriter.create<arith::ConstantOp>(
+      loc, rewriter.getI32IntegerAttr(flags));
+  auto genericMicroKernelOp = rewriter.create<IREE::Codegen::UKernelGenericOp>(
+      loc, outType, fnName, in, out,
+      ValueRange{in_size0, in_size1, in_size2, in_size3, out_size0, out_size1,
+                 flagsVal},
+      /*strided_outer_dims=*/rewriter.getIndexAttr(1));
+  return cast<IREE::Codegen::UKernelOpInterface>(
+      genericMicroKernelOp.getOperation());
+}
+
+namespace {
+
+template <typename OpType>
+struct LowerToUKernelPattern : OpRewritePattern<OpType> {
```

Comment:
If you want to drop the template you can do something https://github.com/llvm/llvm-project/blob/bbc983d33a17b87012c18053a816a8984ab191ca/mlir/test/lib/Dialect/Test/TestPatterns.cpp#L1167 . 

Then the call to `matchDAGForUKernel` would need to be called as

```
TypeSwitch<Operation *, FailureOr<IREE::Codegen::UKernelOpInterface>>(op)
.Case<linalg::MatmulOp, linalg::Mmt4DOp, tensor::PackOp, tensor::UnPackOp>(
    [&](auto op) { return matchDAGForUKernel(rewriter, op); })
.Default([&](Operation *op) {
   return rewriter.notifyMatchFailure(op, "failed to find microkernel op to replace with");
});
```

Reply:
Thanks!